Make initial and max memory sizes be in pages instead of bytes

dschuff opened the pull request:
The AST and everything that uses it treats the values as
pages. Javascript continues to use bytes.
This matches v8 and sexpr-wasm, and the consensus from live discussion
and  PR209 in the spec.

grow_memory hasn't been updated yet, but IIUC it will use pages as well.

dschuff commented:
This will probably require pinned test/revision updates which I haven't done yet.

kripken commented on src/s2wasm.h:
```diff
@@ -55,7 +55,8 @@ class S2WasmBuilder {
         globalBase(globalBase),
-        nextStatic(globalBase) {
+        nextStatic(globalBase),
+        initial_memory(globalBase) {
```
please make it `initialMemory` in camelCase like the others

dschuff replied:
Done.

kripken commented on src/wasm-js.cpp:
```diff
@@ -69,7 +69,7 @@ extern "C" void EMSCRIPTEN_KEEPALIVE load_asm2wasm(char *input) {
   module->memory.initial = EM_ASM_INT_V({
     return Module['providedTotalMemory']; // we receive the size of memory from emscripten
-  });
+  }) / Memory::kPageSize;
```
this and the one after it need to be rounded up, like in `s2wasm`, don't they?

dschuff replied:
Actually it should be validated rather than rounded. It can't be rounded down because then there might not be enough space in linear memory, and it can't be rounded up because the interpreter would think it had more memory than was in the buffer.

JSStats commented:
Might want to take a look at https://github.com/WebAssembly/binaryen/pull/121 and can I close it if your work addresses it all?

dschuff commented:
@JSStats thanks, I didn't see that. I stole a couple of bits from there (although I do prefer the approach of making the page size constant a member of wasm::Memory rather than a preprocessor define).

dschuff commented:
OK, rebased. Alon, do you want to have a look?

kripken commented on check.py:
```diff
@@ -353,7 +353,7 @@ def minify_check(wast, verify_final_result=True):
 
 if len(requested) == 0:
-  BLACKLIST = []
+  BLACKLIST = ['address.wast', 'traps.wast']
```
what's this?

dschuff replied:
That was needed because before https://github.com/WebAssembly/spec/pull/209 those tests had memory in units of bytes. But that landed, so I'll update this patch and this can hopefully go away.

kripken commented on test/s2wasm_known_binaryen_shell_test_failures.txt:
```diff
@@ -28,7 +28,6 @@ bitfld-5.c.s.wast
 memcpy-1.c.s.wast
 mode-dependent-address.c.s.wast
 multi-ix.c.s.wast
-pr29006.c.s.wast
```
what changed here?

dschuff replied:
Don't know? The only relevant change was the size of the memory, so maybe there's an off-by-one error somewhere in the binaryen interpreter  or something that gets papered over because of the larger memory? (or in LLVM, but that seems less likely because this test works in V8)

dschuff commented:
OK, so I rebased and updated the spec tests (which makes `address.wast` and `traps.wast` pass). Now we have the problem that the spec implements rotate and binaryen doesn't. So `i32.wast` and `i64.wast` fail. So we can either leave the spec repo where it is and blacklist address and traps, or update it and blacklist i32 and i64. And then of course we can just implement rotate.

kripken commented:
Unless it's urgent, how about just implementing rotate? I worry about having multiple branches in flight where some disable core spec tests like i32 or address, we could get into confusing situations.

dschuff commented:
So, I actually did start that. I can go ahead and finish it.

dschuff commented:
Rebased, sqashed, added rotates.

kripken commented on src/wasm.h:
```diff
@@ -503,6 +504,32 @@ class Literal {
+  Literal rotR(const Literal& other) const {
+    switch (type) {
+      case WasmType::i32: return Literal(ror(uint32_t(i32), uint32_t(other.i32)));
+      case WasmType::i64: return Literal(ror(uint64_t(i64), uint64_t(other.i64)));
```
everything looks good except I'm puzzled at these `ror` and `rol` methods - where are those provided?

dschuff replied:
rol is defined as a template on like 508, and ror is defined on 521.

dschuff replied:
Actually those should probably private methods, or just free functions. let me fix that.

dschuff replied:
ok, done.

kripken replied:
got it. perhaps let's move these to `src/support/bits.h`? lgtm with that.
